### PR TITLE
blockdev_commit_with_ioerror: commit after io error

### PR DESCRIPTION
**Pull request by aliang123**
Do live commit after block error(read/write/no space)

Signed-off-by: Aihua Liang <aliang@redhat.com>
id:1994948

### CODE REVIEW

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.blockdev_commit_with_ioerror.q35: PASS (98.75 s)

**Comment by aliang123** [comment]
@zhencliu @ngu-niny, please help to review.

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.blockdev_commit_to_nospace.q35: PASS (48.69 s)

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.blockdev_commit_with_ioerror.q35: PASS (97.67 s)

File: qemu/tests/blockdev_commit_with_ioerror.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        """Generate temp data file in guest"""
+        self.session = self.main_vm.wait_for_login()
+        file_path = "%s/%s" % (root_dir, filename)
+        count = int(
+            utils_numeric.normalize_data_size(
+                size,
+                order_magnitude="M",
+                factor=1024))
+        dd_cmd = self.main_vm.params.get(
+            "dd_cmd", "dd if=/dev/urandom of=%s bs=1M count=%s oflag=direct")
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_with_ioerror.py:22]: I think we can use count=1000 directly here, for we are using 1000 in the dd_done msg.

File: qemu/tests/blockdev_commit_with_ioerror.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        mk_file_cmd = dd_cmd % (file_path, count)
+        try:
+            self.session.cmd(mk_file_cmd, timeout=timeout)
+        except ShellTimeoutError as e:
+            if self.params["error_msg"] not in str(e):
+                self.test.fail(str(e))
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_with_ioerror.py:28]: The error_msg check is not required here, for we defined ShellTimeoutError with that error msg, i.e. once ShellTimeoutError is caught, we always get that message

File: qemu/tests/blockdev_commit_with_ioerror.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if not self.session:
+            self.session = self.main_vm.wait_for_login()
+        output = self.session.cmd_output('\n', timeout=120)
+        if self.params["dd_done"] not in output:
+            self.test.fail("dd not continue to run after vm resume")
+        for info in self.files_info:
+            file_path = "%s/%s" % (info[0], info[1])
+            md5_cmd = "md5sum %s > %s.md5 && sync" % (file_path, file_path)
+            self.session.cmd(md5_cmd, timeout=120)
+        self.session.close()
```
Comment (zhencliu) [inline qemu/tests/blockdev_commit_with_ioerror.py:44]: After dd done, we got a md5sum from the original file, after vm reboot, we got another. I think these two are always the same, as we've already covered that test. My point is, do we need to choose a original file to dd? I created a 1G file and got its md5sum before creating snapshot as the original file, then create snapshot, dd if=/the/1G/file of=xxx, we hit io-error, then do commit, dd done, we got another md5sum, compare them.
1G is too large, a 200M file is OK, for you create a snapshot in a 100M space

File: qemu/tests/blockdev_commit_with_ioerror.py (hunk 2 above)
Comment (zhencliu) [reply qemu/tests/blockdev_commit_with_ioerror.py:28]: As per above, error_msg can be removed from cfg

File: qemu/tests/blockdev_commit_with_ioerror.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    """
+    Block commit after io error
+
+    1). create small space(100M)
+    2). start vm with 2G data disk and dd 500M file on it
+    3). create snapshot node on small space
+    4). dd 500M file in guest to casue vm paused with io-error
+    5). do commit from snapshot to base
+    6). continue vm, wait dd finished
+    7). check md5 value of step1 and step6
```
Comment (ngu-niny) [inline qemu/tests/blockdev_commit_with_ioerror.py:103]: It's a confusion with 'step1 and step6', should be some files.

**Comment by ngu-niny** [comment]
Others LGTM.

1-Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_commit_with_ioerror Finshed 2021-08-30 00:22:05 2021-08-30 00:24:39 PASS 154.11158

Acked-by: Nini Gu ngu@redhat.com

File: qemu/tests/blockdev_commit_with_ioerror.py (hunk 4 above)
Comment (aliang123) [reply qemu/tests/blockdev_commit_with_ioerror.py]: Have updated, thanks @ngu-niny

**Comment by zhencliu** [comment]
LGTM

**Pull request merged**